projects/vdk-plugins/vdk-impala/src/vdk/plugin/impala/impala_error_handler.py

```diff
@@ -441,3 +444,78 @@ def _is_pool_error(exception: Exception) -> bool:
+        if errors.exception_matches(
+            exception,
+            classname_with_package="impala.error.OperationalError",
+            exception_message_matcher_regex=".*Memory limit exceeded:.*",
+        ):
+            # We are going to try to increase the memory limits and see if the query passes
+            # But we won't do anything if the sql statement itself sets a memory limit
+            self._log.info(
+                "Query failed with memory error. We are going to increase the memory limit."
+                f"Error was: {exception.__class__}: {str(exception)}"
+            )
+            if (
+                "set memory_limit="
+                in str(
+                    recovery_cursor.get_managed_operation().get_operation_parameters_tuple()
+                ).lower()
+            ):
+                self._log.info(
+                    "The SQL statement you are trying to execute contains a set memory_limit option "
+                    "so we are not going to handle the memory error."
+                )
+                return False
+
+            # Incrementally increase the memory limit and as a last resort try to set the memory to an extreme value
+            if recovery_cursor.get_retries() > 3:
+                #  We won't be able to handle this error by increasing limits
+                return False
+            if recovery_cursor.get_retries() > 2:
+                # An extreme case but let's try it as a last resort
+                recovery_cursor.execute("set memory_limit=512GB;")
+            elif recovery_cursor.get_retries() > 1:
+                self._update_memory_limit(exception, recovery_cursor, 2.0)
+            elif recovery_cursor.get_retries() > 0:
+                self._update_memory_limit(exception, recovery_cursor, 1.5)
+            else:
+                self._update_memory_limit(exception, recovery_cursor, 1.2)
+            # retry the query
+            recovery_cursor.retry_operation()
+            return True
+        else:
+            return False
+
+    def _update_memory_limit(self, exception, recovery_cursor, multiplier):
+        new_memory_limit = self._get_new_memory_limit(
+            exception=exception, multiplier=multiplier
+        )
+        if new_memory_limit:
+            self._log.info(
+                f"Setting memory limit to {new_memory_limit} bytes and retrying SQL statement."
+            )
+            recovery_cursor.execute(f"set mem_limit={new_memory_limit};")
+        else:
+            self._log.warning(
+                f"Unable to determine current memory limit for statement."
+            )
+
+    @staticmethod
+    def _convert_to_bytes(value: str, unit: str):
+        units = {"KB": 1024, "MB": 1024**2, "GB": 1024**3, "TB": 1024**4}
+        return int(float(value) * units.get(unit, 1))
+
+    def _get_new_memory_limit(self, exception, multiplier: float):
+        memory_limit_match = re.search(MEMORY_LIMIT_PATTERN, str(exception))
```

Comment:
In my experience, (far from big providedly) parsing is one of the biggest source of bugs. I always try to make sure I follow a spec or a contract or something stable. 

In this case Impala is printing this in https://github.com/apache/impala/blob/master/be/src/runtime/mem-tracker.cc#L314 

using PrettyPrinter whose implementation is [here](https://github.com/apache/impala/blob/ae14d78c8f2e8366e67aa8c39f0c02c60862905e/be/src/util/pretty-printer.h#L127) 

And from git blame I can see that code has not changed for 6-9 years. So we are probably ok. 


Also the pattern likely won't catch false positive. At least ([searching in the code](https://github.com/search?q=repo%3Aapache%2Fimpala+%22Limit%3D%22&type=code&p=5) that seems the only "Limit=" printed

Reply:
I based the logic on several exceptions I reviewed from our internal system and this works best.